### PR TITLE
Add user autoprovisioning via libreGraph

### PR DESCRIPTION
**Pull request by rhafer**
## Description

When removing the accounts service we lost the user autoprovision feature. This re-introduces it. When autoprovisioning is enabled (via PROXY_AUTOPROVISION_ACCOUNTS, as in the past) accounts that are not resolvable via cs3 will be provsioned via the libregraph API.

## Related Issue:
- Fixes https://github.com/owncloud/ocis/issues/3540

## How Has This Been Tested?

- test environment:
- test case 1:
- test case 2:
- ...

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

**Comment by mmattel** [comment]
If applicapable, in case "new" env's are coming in, do not forget to give them a description and a default.

File: extensions/graph/pkg/middleware/auth.go [hunk 1]
```diff
@@ -25,6 +27,8 @@ func authOptions(opts ...account.Option) account.Options {
 // Auth provides a middleware to authenticate requests using the x-access-token header value
 // and write it to the context. If there is no x-access-token the middleware prevents access and renders a json document.
 func Auth(opts ...account.Option) func(http.Handler) http.Handler {
+	// Note: This largely duplicates was ocis-pkg/middleware/account.go already does (apart from a slightly different error
```
Comment (C0rby) [inline extensions/graph/pkg/middleware/auth.go:30]:
```suggestion
	// Note: This largely duplicates what ocis-pkg/middleware/account.go already does (apart from a slightly different error
```

**Review by C0rby — APPROVED** [review]

**Review by kobergj — CHANGES_REQUESTED** [review]
Found some points, but looks good in general!

File: extensions/proxy/pkg/user/backend/cs3.go [hunk 2]
```diff
@@ -114,9 +133,190 @@ func (c *cs3backend) Authenticate(ctx context.Context, username string, password
...
+	return cs3user
+}
+
+var autoProvisionUserCreator *cs3.User
+
+// This returns an hardcoded internal User, that is privileged to create new User via
+// the Graph API. This user is needed of autoprovisioning of user from incoming OIDC
+// claims.
+func getAutoProvisionUserCreator() (*cs3.User, error) {
+	if autoProvisionUserCreator == nil {
```
[the remark this answers is not in the document]
Comment (rhafer) [reply extensions/proxy/pkg/user/backend/cs3.go:281]: Hm, I guess I'll just rework this to always generate the user. It's not in a critical path at all. ;-)

File: extensions/proxy/pkg/user/backend/cs3.go [hunk 3]
```diff
@@ -114,9 +133,190 @@ func (c *cs3backend) Authenticate(ctx context.Context, username string, password
 }
 
 func (c *cs3backend) CreateUserFromClaims(ctx context.Context, claims map[string]interface{}) (*cs3.User, error) {
```
[the remark this answers is not in the document]
Comment (rhafer) [reply extensions/proxy/pkg/user/backend/cs3.go:139]: Done

**Pull request merged**